### ISSUE TITLE
interactive node connection closed by remote host

### ISSUE THREAD

**Issue by maituoy** [ArjunaCluster/ArjunaUsers#71]
If you notice an issue on arjuna, please first search the existing issues and ensure that it has not been reported. If you notice a similar example, please comment on that issue.

> Please update the issue's title with something informative

Please provide the following information to help us help you:

## Basic Info
Your Name: Maituo Yu
Your Andrew ID: maituoy

## Where it happened
Job Ids: 8427
Node(s) on which the problem occurred: e001 (But on other nodes as well)

## What Happened
Expected Behavior: The node will keep connected when running a interactive node
Observed Behavior: The connection to the granted node was closed by remote host after a certain amount of time. 

## Log Files
Location of Log file Showing the Error: No log file
Location of Script showing [Minimum Working Example]: ```/home/maituoy/software/interdispipe/example/Ni2MnIn/HSE/band```

Please also attach any logs and the submission script to this issue.

```log # 'log' here because it's a log file
Submit log files like this (Notice the ``` before and after this block)
this ensures logs are formated correctly (Ie. Comments don't become headings)

```

```bash # 'bash' here it's a bash script
salloc -N 1 -n 54 -p cpu --mem-per-cpu=0 --time=7-00:00:00

```

[Minimum Working Example]: https://en.wikipedia.org/wiki/Minimal_working_example

## What I've Tried

Please List what you've tried to debug the issue. Please include commands and resulting output

1) I googled but nothing have been found similar so far.

```shell # 'shell' here because we're showing a shell command
> awesome_command
Output from awesome_command
```

> If you do not have any of the above, please explain why you do not have it, and submit the issue, however, the more information you give us, the better we can help you.

Screenshot showing the disconnection.
<img width="481" alt="image" src="https://user-images.githubusercontent.com/37908868/134792541-8bca6676-5e9f-4a1c-9b67-491e324d11fa.png">

**Comment by awadell1** [comment]
My guess is you're doing something like:

```
salloc ....
ssh e001
```

The problem is `ssh e001` is outside of slurm's control, making it an unauthorized process, and thus it gets terminated. This is a new feature to ensure jobs get the resources requested and automatically clean up wayward jobs. The check happens on an hourly basis which is why your terminal doesn't get killed right away.

I would recommend switching to something like: `srun -N 1 -n 54 -p cpu --pty bash`.

Looking at e001's logs I'm guessing the disconnect happened around `20210925 17:39:27`. Can you confirm?

**Comment by maituoy** [comment]
Yeah, that's exactly what I was doing. I'm trying ```srun -N 1 -n 54 -p cpu --pty bash```. But I'm not able to run executables with ```mpirun```. It stuck at some point and nothing was output. I do it in ```mpirun -np 54 vasp_ncl > vasp.out```. I check the ```mpirun``` module and I think it's well loaded. The single core calculation is not affected.
<img width="496" alt="image" src="https://user-images.githubusercontent.com/37908868/134818867-7a75ec39-6245-46fc-8ab7-db811c518914.png">

**Comment by awadell1** [comment]
Maybe try: https://github.com/ArjunaCluster/ArjunaUsers/issues/59#issuecomment-920091891

**Comment by maituoy** [comment]
You mean run like ```srun -n 54 vasp_ncl > vasp.out```?

**Comment by awadell1** [comment]
Roughly. I think he was running in an sbatch allocation so maybe not exactly the interactive environment you want. And idk how srun within srun behaves.

**Comment by maituoy** [comment]
No, it does not work. Job is terminated after a while. 
<img width="639" alt="image" src="https://user-images.githubusercontent.com/37908868/134821116-c33ae0f7-f84f-4143-82b4-e1bb8fa8e80d.png">

**Comment by awadell1** [comment]
Did you also install vasp linked to `openmpi schedulers=slurm pmi=true`? But I would try reducing your MWE and trying to isolate the problem. My openmpi / vasp knowledge is pretty limited geneally.

Definitely don't try to start an interactive job with `salloc` and then sshing in, slurm will kill your processes. If you do find a solid way to do an interactive mpi job, we'd love PR to update the docs (#4).

**Comment by maituoy** [comment]
I did not use the Spack to install VASP since it was not able to find the local package mirror.

**Comment by maituoy** [comment]
I'm not sure if this is due to the Slurm version since I see a similar issue ([mpirun hang](https://github.com/open-mpi/ompi/issues/8378)) posted in the open-mpi git repo. I have checked that we are currently using the ```slurm 20.11.8``` and the latest version is ```slurm 21.08.0```. Is it possible to upgrade the slurm without affecting others' running job?

**Comment by awadell1** [comment]
No and slurm will not be updated any time soon. Slurm 21.08.0 was released on August 26th right at the tail end of the upgrade. Previously we were running a roughly 5-year-old version.

The linked issue suggests that 21.11.03 has the fixed so I don't think that's your issue
https://github.com/open-mpi/ompi/issues/8378#issuecomment-761223699

**Comment by maituoy** [comment]
> Did you also install vasp linked to `openmpi schedulers=slurm pmi=true`? But I would try reducing your MWE and trying to isolate the problem. My openmpi / vasp knowledge is pretty limited geneally.
> 
> Definitely don't try to start an interactive job with `salloc` and then sshing in, slurm will kill your processes. If you do find a solid way to do an interactive mpi job, we'd love PR to update the docs (#4).

It does not have to be the VASP, since the same issue exists for other software as well like Quantum Espresso. I think it's more like a system issue, not the software.

**Comment by awadell1** [comment]
We have encountered some OpenMPI growing pains and would appreciate a minimal working example of this issue. The system OpenMPI is not Slurm aware, and that does seem to be causing problems.

Again have you tried following #59 and installing a slurm-aware openmpi? There have been reports of this working for py-gpaw. 

To clarify, if you ssh into a node that you have an salloc (or any other job) on, **the expected behavior is for that terminal (and all related processes) to be killed by slurm**. This is to ensure users get the resources they request and will _not_ be changing.

**Comment by awadell1** [comment]
Correction: the System OpenMPI is slurm aware. Apologies for the confusion

**Comment by maituoy** [comment]
Unfortunately, none of the scientific packages we usually use are able to be installed via spack. That means I was not able to test a package linked to OpenMPI and slurm. The good news is I found out that we can run the parallel jobs on the interactive nodes by not login into the nodes. Basically, I just submit an interactive job using the ```salloc``` and run executable binaries with ```srun``` right after allocation is granted.  I'm still testing this approach. So far, it works well if your job only involves a scientific package. However, if you would like to embed your calculation in the subprocess of python script (such as using an ASE calculator), it can easily encounter a memory issue on complicated calculations (I compared the PBE and HSE functional in DFT).

**Comment by maituoy** [comment]
BTW, I don't think #59 was using the interactive nodes. The regular Slurm jobs work for us as well.

**Comment by maituoy** [comment]
Sorry, do you mean I will be able to log into the granted node via  ssh again?

**Comment by awadell1** [comment]
You can do:
```shell
awadell in coe in ~
❯ salloc -c 4 --mem 4G -p cpu
salloc: error: Ignoring conflicting option "x11" in plugin "x11"
salloc: Granted job allocation 19201
salloc: Waiting for resource configuration
salloc: Nodes d018 are ready for job

awadell in coe in ~
❯ srun --pty bash
srun: error: Ignoring conflicting option "x11" in plugin "x11"
slurmstepd: error: couldn't chdir to `/opt/ArjunaAdmin': No such file or directory: going to /tmp instead
slurmstepd: error: couldn't chdir to `/opt/ArjunaAdmin': No such file or directory: going to /tmp instead

awadell in d018 in ~
❯ awesome_commands
```

Or:

```shell
awadell in coe in ~
❯ srun -c 4 --mem 4G -p cpu --pty bash
srun: error: Ignoring conflicting option "x11" in plugin "x11"

awadell in d018 in ~
❯ awesome_commands
```

You may not:
```shell
❯ salloc -c 4 --mem 4G -p cpu
salloc: error: Ignoring conflicting option "x11" in plugin "x11"
salloc: Granted job allocation 19205
salloc: Waiting for resource configuration
salloc: Nodes d018 are ready for job

awadell in coe in ~
❯ ssh d018

awadell in d018 in ~
❯ this_ssh_session_will_be_killed
```

**Comment by maituoy** [comment]
I'm still not able to run with mpirun. I used exactly the same code as yours.
When I tried to run mpirun with the OpenMPI, an error message was given as below:
<img width="560" alt="image" src="https://user-images.githubusercontent.com/37908868/138013207-51db4dcd-c2bd-422d-87dc-a98a98084fa9.png">
I tried the ```--oversubscribe``` option but it just ran with 4 jobs and each one uses 1 core. And the ```--use-hwthread-cpus``` does not work at all.

And if I run with the intel mpi, the mpirun will just hang there.
